kafka/consumer.go

```diff
@@ -203,9 +227,16 @@ func (c *Consumer) fetch(ctx context.Context) error {
 		errors.Is(fetches.Err0(), context.DeadlineExceeded) {
 		return context.Canceled
 	}
-	// Acquire the lock after checking if the the context is cancelled and/or
-	// the client is closed.
-	c.mu.RLock()
+	// Acquire the lock after checking if the the context is done.
+	// If the lock cannot be acquired, the consumer is closing,
+	// be closing, in which case, return with context.Cancelled.
+	if !c.mu.TryRLock() {
```

Comment:
Not a fan of `TryRLock` usage. WDYT about interrupting the `for` loop in `Run` if `c.closed` is closed and in the `Close` method, waiting for `Run` to finish before proceeding to close stuff. Example (only for conveying the intention):

```go
func (c *Consumer) Run(ctx context.Context) error {
	c.mu.Lock()
	select {
	case <-c.running:
		c.mu.Unlock()
		return apmqueue.ErrConsumerAlreadyRunning
	default:
		close(c.running)
	}
	c.mu.Unlock()
        defer close(c.runStopped)
	for {
                select {
                case <-c.closed:
                    return nil // or ErrClosed
                default:
                }
		if err := c.fetch(ctx); err != nil {
			if errors.Is(err, context.Canceled) {
				return nil // Return no error if err == context.Canceled.
			}
			return err
		}
	}
}
```

Then in `Close`:

```go
func (c *Consumer) Close() error {
	c.mu.Lock()
	select {
	case <-c.closed:
	default:
		close(c.closed)
        }
        c.mu.Unlock()
        <-c.runStopped // can choose to pass a context to Close and wait with select

        c.mu.Lock()
        defer c.mu.Unlock()
        // Close all partition consumers first to ensure that there aren't
        // any records being processed while the kgo.Client is being closed.
        // Also ensures that commits can be issued after the records are
        // processed when AtLeastOnceDelivery is configured.
        c.consumer.close() // if this requires to be done only once, we can use a bool flag or maybe set the values to nil if successfully closed
        c.client.Close()
	c.consumer.wg.Wait() // Wait for consumer goroutines to exit.
	return nil
}
```